src/TADA_DataRetrieval_new.R

```diff
@@ -0,0 +1,369 @@
+  }
+  
+  if (length(organization) > 1) {
+    WQPquery <- c(WQPquery, organization = list(organization))
+  } else if (organization != "null") {
+    WQPquery <- c(WQPquery, organization = organization)
+  }
+  
+  if (length(endDate) > 1) {
+    if (is.na(suppressWarnings(lubridate::parse_date_time(endDate[1], orders = "ymd")))) {
+      stop("Incorrect date format. Please use the format YYYY-MM-DD.")
+    }
+    WQPquery <- c(WQPquery, endDate = list(endDate))
+  } else if (endDate != "null") {
+    if (is.na(suppressWarnings(lubridate::parse_date_time(endDate, orders = "ymd")))) {
+      stop("Incorrect date format. Please use the format YYYY-MM-DD.")
+    }
+    WQPquery <- c(WQPquery, endDate = endDate)
+  }
+  
+  # Retrieve all 3 profiles
+  print("Downloading WQP query results. This may take some time depending upon the query size.")
+  results.DR <- dataRetrieval::readWQPdata(WQPquery,
+                                           dataProfile = "resultPhysChem",
+                                           ignore_attributes = TRUE
+  )
+  # check if any results are available
+  if ((nrow(results.DR) > 0) == FALSE) {
+    print("Returning empty results dataframe: Your WQP query returned no results (no data available). Try a different query. Removing some of your query filters OR broadening your search area may help.")
+    TADAprofile.clean <- results.DR
+  } else {
+    sites.DR <- dataRetrieval::whatWQPsites(WQPquery)
+    
+    projects.DR <- dataRetrieval::readWQPdata(WQPquery,
+                                              ignore_attributes = TRUE,
+                                              service = "Project"
+    )
+    
+    TADAprofile <- TADA_JoinWQPProfiles(
+      FullPhysChem = results.DR,
+      Sites = sites.DR,
+      Projects = projects.DR
+    )
+    
+    # need to specify this or throws error when trying to bind rows. Temporary fix for larger
+    # issue where data structure for all columns should be specified.
+    cols <- names(TADAprofile)
+    
+    TADAprofile <- TADAprofile %>% dplyr::mutate_at(cols, as.character)
+    
+    # run TADA_AutoClean function
+    if (applyautoclean == TRUE) {
+      print("Data successfully downloaded. Running TADA_AutoClean function.")
+      
+      TADAprofile.clean <- TADA_AutoClean(TADAprofile)
+    } else {
+      TADAprofile.clean <- TADAprofile
+    }
+  }
+  
+  # subset original dataframe to only those WQP observations within the user-supplied sf:
+  if (!is.null(sf) & inherits(sf, "sf")) {
+    try(TADAprofile.clean <- TADAprofile.clean %>%
+          # transform df into spatial object using a new TADA_MakeSpatial() function
+          TADA_MakeSpatial(data = ., crs = 4326) %>%
+          # transform to the CRS of the sf object
+          sf::st_transform(sf::st_crs(sf)) %>%
+          # select only WQP observations within the sf object
+          .[sf, ] %>%
+          sf::st_drop_geometry(),
+        silent = TRUE)
+  }
```

Comment:
Would it be possible (or useful?) to directly submit spatial data in the original query instead of using it to subset after downloading? That's what I was anticipating this function would do at least, but I'm not sure if it is the right thing to do. For example: https://github.com/DOI-USGS/ds-pipelines-targets-example-wqp/blob/e796fc7a792986865f6f77e184f18986073ca8ea/1_inventory/src/get_wqp_inventory.R#L61-L66

Reply:
I think that could definitely speed things up!